fs/vfs/fs_poll.c

```diff
@@ -284,24 +285,28 @@ void poll_notify(FAR struct pollfd **afds, int nfds, pollevent_t eventset)
       fds = afds[i];
       if (fds != NULL)
         {
-          /* race condition protection when modifying fds->revents */
-
-          flags = enter_critical_section();
-
           /* The error event must be set in fds->revents */
 
-          fds->revents |= eventset & (fds->events | POLLERR | POLLHUP);
-          if ((fds->revents & (POLLERR | POLLHUP)) != 0)
+          revents = eventset & (fds->events | POLLERR | POLLHUP);
+          if (revents != 0)
             {
-              /* Error or Hung up, clear POLLOUT event */
+              /* race condition protection when modifying fds->revents */
 
-              fds->revents &= ~POLLOUT;
-            }
+              flags = enter_critical_section();
+
+              fds->revents |= revents;
+              if ((fds->revents & (POLLERR | POLLHUP)) != 0)
+                {
+                  /* Error or Hung up, clear POLLOUT event */
+
+                  fds->revents &= ~POLLOUT;
+                }
 
-          leave_critical_section(flags);
+              leave_critical_section(flags);
+            }
 
-          if ((fds->revents != 0 || (fds->events & POLLALWAYS) != 0) &&
-              fds->cb != NULL)
+          if (fds->cb != NULL &&
+              (revents != 0 || (fds->events & POLLALWAYS))
```

Comment:
If section above might modify fds->revents value so it's not anymore same as local revents

Reply:
Yes. So you don't want to call the callback for any event that was not notified for this call, so checking fds->revents is wrong here in any case. But, if there is indeed a case that someone calls this with POLLERR and POLLOUT or POLLHUP and POLLOUT, then the POLLOUT should not be counted in. I don't think that makes much sense though (?) I can of course clear out the POLLOUT flag also from revents.